File: python/0239.py
# Link: https://leetcode.com/problems/sliding-window-maximum/description/
# Runtime: 1334 ms
# Memory: 33.38 MB

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        potential_maximums = deque()
        for i in range(k):
            self.__remove_redundant_values(nums, potential_maximums, nums[i])
            potential_maximums.append(i)
        
        maximums = []
        maximums.append(nums[potential_maximums[0]])
        for i in range(k, len(nums)):
            if potential_maximums[0] == i - k: potential_maximums.popleft()
            self.__remove_redundant_values(nums, potential_maximums, nums[i])

            potential_maximums.append(i)
            maximums.append(nums[potential_maximums[0]])

        return maximums
    
    def __remove_redundant_values(self, nums: List[int], potential_maximums: Deque[int], current_value: int) -> None:
        while (len(potential_maximums) > 0) and (nums[potential_maximums[-1]] <= current_value):
                potential_maximums.pop()

File: c++/0239.cpp
// Link: https://leetcode.com/problems/sliding-window-maximum/description/
// Runtime: 193 ms
// Memory: 134.66 MB

class Solution {
public:
  std::vector<int> maxSlidingWindow(const std::vector<int> &nums, int k) {
    std::deque<int> potentialMaximums;
    for (int i = 0; i < k; i++) {
      removeRedundantValues(nums, potentialMaximums, nums[i]);

      potentialMaximums.push_back(i);
    }

    std::vector<int> maximums;
    maximums.push_back(nums[potentialMaximums.front()]);
    for (int i = k; i < static_cast<int>(nums.size()); i++) {
      if (potentialMaximums.front() == i - k)
        potentialMaximums.pop_front();
      removeRedundantValues(nums, potentialMaximums, nums[i]);

      potentialMaximums.push_back(i);
      maximums.push_back(nums[potentialMaximums.front()]);
    }

    return maximums;
  }

private:
  void removeRedundantValues(const std::vector<int> &nums,
                             std::deque<int> &potentialMaximums,
                             int currentValue) {
    while ((!potentialMaximums.empty()) &&
           (nums[potentialMaximums.back()] <= currentValue))
      potentialMaximums.pop_back();
  }
};
